dbms/src/Flash/Coprocessor/DAGExpressionAnalyzerHelper.cpp

```diff
@@ -280,6 +281,102 @@ String DAGExpressionAnalyzerHelper::buildCastFunction(
+    auto * function_build_ptr = function_builder.get();
+    if (auto * default_function_builder = dynamic_cast<DefaultFunctionBuilder *>(function_build_ptr);
+        default_function_builder)
+    {
+        auto * function_impl = default_function_builder->getFunctionImpl().get();
+        if (auto * function_json_unquote = dynamic_cast<FunctionsJsonUnquote *>(function_impl); function_json_unquote)
+        {
+            function_json_unquote->setNeedValidCheck(valid_check);
+        }
+        else
+        {
+            throw Exception(fmt::format("Unexpected func {} in buildJsonUnquoteFunction", func_name));
+        }
+    }
+    else
+    {
+        throw Exception(fmt::format("Unexpected func {} in buildJsonUnquoteFunction", func_name));
+    }
+
+    const ExpressionAction & apply_function
+        = ExpressionAction::applyFunction(function_builder, {arg}, result_name, nullptr);
+    actions->add(apply_function);
+    return result_name;
+}
+
+String DAGExpressionAnalyzerHelper::buildCastJsonAsStringFunction(
+    DAGExpressionAnalyzer * analyzer,
+    const tipb::Expr & expr,
+    const ExpressionActionsPtr & actions)
+{
+    if unlikely (expr.children_size() != 1)
+        throw TiFlashException("CAST function only support one argument", Errors::Coprocessor::BadRequest);
+    if unlikely (!exprHasValidFieldType(expr))
+        throw TiFlashException("CAST function without valid field type", Errors::Coprocessor::BadRequest);
+
+    const auto & input_expr = expr.children(0);
+    auto func_name = getFunctionName(expr);
+
+    String arg = analyzer->getActions(input_expr, actions);
+    const auto & collator = getCollatorFromExpr(expr);
+    String result_name = genFuncString(func_name, {arg}, {collator});
+    if (actions->getSampleBlock().has(result_name))
+        return result_name;
+
+    const FunctionBuilderPtr & function_builder = FunctionFactory::instance().get(func_name, analyzer->getContext());
+    auto * function_build_ptr = function_builder.get();
+    if (auto * default_function_builder = dynamic_cast<DefaultFunctionBuilder *>(function_build_ptr);
+        default_function_builder)
+    {
+        auto * function_impl = default_function_builder->getFunctionImpl().get();
+        if (auto * function_cast_json_as_string = dynamic_cast<FunctionsCastJsonAsString *>(function_impl);
+            function_cast_json_as_string)
+        {
+            function_cast_json_as_string->setOutputTiDBFieldType(expr.field_type());
+        }
+        else
+        {
+            throw Exception(fmt::format("Unexpected func {} in buildCastJsonAsStringFunction", func_name));
+        }
+    }
+    else
+    {
+        throw Exception(fmt::format("Unexpected func {} in buildCastJsonAsStringFunction", func_name));
+    }
+
+    const ExpressionAction & apply_function
+        = ExpressionAction::applyFunction(function_builder, {arg}, result_name, nullptr);
+    actions->add(apply_function);
+    return result_name;
```

Comment:
It looks like they can be combined into one function

Reply:
Do you mean extracting some common code here?

Reply:
I mean that refine the code like
```
String DAGExpressionAnalyzerHelper::buildJsonFunction(
    DAGExpressionAnalyzer * analyzer,
    const tipb::Expr & expr,
    const ExpressionActionsPtr & actions)
{
    if unlikely (!exprHasValidFieldType(expr))
        throw TiFlashException("JsonUnquote function without valid field type", Errors::Coprocessor::BadRequest);

    const auto & input_expr = expr.children(0);
    auto func_name = getFunctionName(expr);

    String arg = analyzer->getActions(input_expr, actions);
    const auto & collator = getCollatorFromExpr(expr);
    String result_name = genFuncString(func_name, {arg}, {collator});
    if (actions->getSampleBlock().has(result_name))
        return result_name;

    const FunctionBuilderPtr & function_builder = FunctionFactory::instance().get(func_name, analyzer->getContext());
    auto * function_build_ptr = function_builder.get();
    if (auto * default_function_builder = dynamic_cast<DefaultFunctionBuilder *>(function_build_ptr);
        default_function_builder)
    {
        auto * function_impl = default_function_builder->getFunctionImpl().get();
        if (auto * function_cast_json_as_string = dynamic_cast<FunctionsCastJsonAsString *>(function_impl);
            function_cast_json_as_string)
        {
            function_cast_json_as_string->setOutputTiDBFieldType(expr.field_type());
        }
        else if (auto * function_json_unquote = dynamic_cast<FunctionsJsonUnquote *>(function_impl); function_json_unquote)
        {
const auto & input_expr = expr.children(0);
            bool valid_check = !(isScalarFunctionExpr(input_expr) && input_expr.sig() == tipb::ScalarFuncSig::CastJsonAsString);
            function_json_unquote->setNeedValidCheck(valid_check);
        }
        else
        {
            throw Exception(fmt::format("Unexpected func {}", func_name));
        }
    }
    else
    {
        throw Exception(fmt::format("Unexpected func {}", func_name));
    }

    const ExpressionAction & apply_function
        = ExpressionAction::applyFunction(function_builder, {arg}, result_name, nullptr);
    actions->add(apply_function);
    return result_name;
}
```

Reply:
How about merging with https://github.com/pingcap/tiflash/pull/8333/files#diff-e78d988899b5d199db66439fd853d0c86526e77cbffcb20c2c9d10500bdb51eaR284